### PR TITLE
GLP Cauldron

### PR DESCRIPTION
**Pull request by 0xCalibur**
- [x] CauldronV4
- [x] DegenBoxOwner
- [x] GLP Oracle
- [x] GLP Cauldron
- [x] GLP Wrapper
- [x] Simple Mim Distributor for Cauldron Repay
- [x] Test esGMX rewards

### CODE REVIEW

**Comment by 0xm3rlin** [comment]
One thing in Cauldron V4, to allow the functionality in https://github.com/Abracadabra-money/abracadabra-money-foundry/blob/97b1f6b1877430dff5cbf670a26d4473bbe143d6/src/cauldrons/CauldronV4.sol#L499, we should make sure to blacklist the bentobox owner caller in the mastercontract, leaving the comment to not forget as much

File: src/cauldrons/CauldronV4.sol [hunk 1]
```diff
@@ -415,7 +415,13 @@ contract CauldronV4 is BoringOwnable, IMasterContract {
             callData = abi.encodePacked(callData, value1, value2);
         }
 
-        require(callee != address(bentoBox) && callee != address(this) && !blacklistedCallees[callee], "Cauldron: can't call");
+        require(
+            callee != address(bentoBox) &&
+                callee != address(this) &&
+                callee != BoringOwnable(address(bentoBox)).owner() &&
+                !blacklistedCallees[callee],
```
Comment (0xm3rlin) [inline src/cauldrons/CauldronV4.sol]: why don't we just use blacklistedCallees here, to forbid the bentoBox.owner() call

File: src/DegenBox.sol [hunk 2]
```diff
@@ -55,6 +56,8 @@ contract MasterContractManager is BoringOwnable, BoringFactory {
     // solhint-disable-next-line var-name-mixedcase
     uint256 private immutable DOMAIN_SEPARATOR_CHAIN_ID;
 
+    mapping(IERC20 => bool) allowedBrokenTotalSupplyTokens;
```
Comment (0xm3rlin) [inline src/DegenBox.sol]: what is this exactly?

File: src/periphery/DegenBoxOwner.sol [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import "interfaces/IBentoBoxV1.sol";
+
+contract DegenBoxOwner is BoringOwnable, IBentoBoxOwner {
+    error ErrNotOperator(address operator);
+
+    event LogOperatorChanged(address indexed operator, bool previous, bool current);
+    event LogDegenBoxChanged(IBentoBoxV1 indexed previous, IBentoBoxV1 indexed current);
+
+    IBentoBoxV1 public degenBox;
+    mapping(address => bool) public strategyReblancers;
```
Comment (0xm3rlin) [inline src/periphery/DegenBoxOwner.sol]: small typo  :)

File: src/DegenBox.sol [hunk 4]
```diff
@@ -319,7 +322,7 @@ contract DegenBox is MasterContractManager, BoringBatchable {
         Rebase memory total = totals[token];
 
         // If a new token gets added, the tokenSupply call checks that this is a deployed contract. Needed for security.
-        require(total.elastic != 0 || token.totalSupply() > 0, "BentoBox: No tokens");
+        require(total.elastic != 0 || (token.totalSupply() > 0 || allowedBrokenTotalSupplyTokens[token]), "BentoBox: No tokens");
```
Comment (0xm3rlin) [inline src/DegenBox.sol]: is this necessary now that we are working with the working staked GLP?

File: src/periphery/GmxGlpRewardHandler.sol [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+    /// @notice claims vested gmx tokens to feeRecipient
+    /// @dev vested esGmx gets converted to GMX every second, so whatever amount is vested gets claimed
+    function claimVestedGmx() external onlyOwner {
+        // stops vesting and stakes the remaining esGMX
+        // this can be used in case glp withdraws are hampered
+        uint256 gmxClaimed = IGmxVester(rewardRouter.glpVester()).claim();
+
+        //Transfer all of the gmx received to fee recipient
+        IERC20(rewardRouter.gmx()).safeTransfer(feeCollector, gmxClaimed);
```
Comment (0xm3rlin) [inline src/periphery/GmxGlpRewardHandler.sol]: maybe instead of transferring it to the fee distributor it could stay here to increase earnings?
we want to use the multiplier if possible.

File: src/periphery/MimRewardsDistributor.sol [hunk 6]
```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+import "BoringSolidity/BoringOwnable.sol";
+import "BoringSolidity/ERC20.sol";
+import "BoringSolidity/libraries/BoringERC20.sol";
+
+/// @notice basic simple mim reward distributor
+/// To be improved to support per-address mim distribution allocations
+contract MimRewardDistributor is BoringOwnable {
```
Comment (0xm3rlin) [inline src/periphery/MimRewardsDistributor.sol]: this does not work yet

File: src/periphery/MimRewardsDistributor.sol [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        _;
+    }
+
+    constructor(ERC20 _mim, address _recipient) {
+        mim = _mim;
+        recipient = _recipient;
+    }
+
+    function distribute() external notPaused {
+        mim.transfer(recipient, mim.balanceOf(address(this)));
```
Comment (0xm3rlin) [inline src/periphery/MimRewardsDistributor.sol]: distribution means repaying on the cauldron.
e.g. repayAll on cauldron v4

File: src/tokens/ERC20Vault.sol [hunk 8]
```diff
@@ -2,16 +2,15 @@
...
 import "interfaces/IERC20Vault.sol";
 
 /// @notice A token that behaves like SushiBar where the contract underlying token balance
 /// influences the share value.
-contract ERC20Vault is IERC20Vault, ERC20WithSupply, BoringOwnable {
+contract ERC20Vault is IERC20Vault, ERC20WithSupply {
     using BoringERC20 for IERC20;
 
-    IERC20 public immutable underlying;
+    IERC20 public underlying;
```
Comment (0xm3rlin) [inline src/tokens/ERC20Vault.sol]: why was this change necessary if not used for wrapper?

File: src/tokens/GmxGlpWrapper.sol [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    function toAmount(uint256 shares) public view returns (uint256) {
+        uint256 totalsGlp = sGlp.balanceOf(address(this));
+        return (totalSupply == 0 || totalsGlp == 0) ? shares : (shares * totalsGlp) / totalSupply;
+    }
+
+    function toShares(uint256 amount) public view returns (uint256) {
+        uint256 totalsGlp = sGlp.balanceOf(address(this));
+        return (totalSupply == 0 || totalsGlp == 0) ? amount : (amount * totalSupply) / totalsGlp;
```
Comment (0xm3rlin) [inline src/tokens/GmxGlpWrapper.sol]: still allowing collateral increases for future extensibility?
i would assume this to be its own cauldron

File: src/DegenBox.sol (hunk 2 above)
Comment (0xm3rlin) [reply src/DegenBox.sol]: is this necessary now that we are working with the working staked GLP?

**Review by 0xm3rlin — CHANGES_REQUESTED** [review]

File: src/DegenBox.sol (hunk 2 above)
Comment (0xCalibur) [reply src/DegenBox.sol]: I forgot to remove it

File: src/periphery/GmxGlpRewardHandler.sol (hunk 5 above)
Comment (0xCalibur) [reply src/periphery/GmxGlpRewardHandler.sol]: the handleRewards already stake and compound to increase the multiplier etc. This call should when we want to exit.

File: src/periphery/MimRewardsDistributor.sol (hunk 6 above)
Comment (0xCalibur) [reply src/periphery/MimRewardsDistributor.sol]: Yes, I'm making a working version to repayment

File: src/periphery/MimRewardsDistributor.sol (hunk 7 above)
Comment (0xCalibur) [reply src/periphery/MimRewardsDistributor.sol]: That was a quick example, it should transfer to cauldron and skim repay

File: src/tokens/ERC20Vault.sol (hunk 8 above)
Comment (0xCalibur) [reply src/tokens/ERC20Vault.sol:13]: If GMX changes the stakedGlp for an updated version or bug fixed (like they did in the past), we can change it here without deprecating the wrapper and the cauldron

File: src/tokens/GmxGlpWrapper.sol (hunk 9 above)
Comment (0xCalibur) [reply src/tokens/GmxGlpWrapper.sol]: I can remove all the share logic and only mint burn 1:1, sounds good?

File: src/tokens/GmxGlpWrapper.sol [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        rewardHandler = _rewardHandler;
+    }
+
+    function setStakedGlp(IERC20 _sGlp) external onlyOwner {
+        emit LogStakedGlpChanged(sGlp, _sGlp);
+        sGlp = _sGlp;
+    }
+
+    // Forward unknown function calls to the reward handler.
+    fallback() external {
```
Comment (0xm3rlin) [inline src/tokens/GmxGlpWrapper.sol:104]: should this be payable btw?, i think the proxy is neat

Comment (0xm3rlin) [reply src/tokens/GmxGlpWrapper.sol:104]: i guess should be fine as it should still call proxy with data

Comment (0xCalibur) [reply src/tokens/GmxGlpWrapper.sol:104]: I don't think we should accept eth in there, do you have any idea in mind if so?

Comment (0xm3rlin) [reply src/tokens/GmxGlpWrapper.sol:104]: should be fine

File: src/tokens/GmxGlpWrapper.sol (hunk 9 above)
Comment (0xm3rlin) [reply src/tokens/GmxGlpWrapper.sol]: maybe let's just keep it simple here.

File: src/tokens/ERC20Vault.sol (hunk 8 above)
Comment (0xCalibur) [reply src/tokens/ERC20Vault.sol]: good catch, forgot to put it back to immutable.

File: src/cauldrons/CauldronV4.sol [hunk 11]
```diff
@@ -415,7 +420,7 @@ contract CauldronV4 is BoringOwnable, IMasterContract {
             callData = abi.encodePacked(callData, value1, value2);
         }
 
-        require(callee != address(bentoBox) && callee != address(this) && !blacklistedCallees[callee], "Cauldron: can't call");
+        require(!blacklistedCallees[callee] && !masterContract.blacklistedCallees(callee), "Cauldron: can't call");
```
Comment (0xm3rlin) [inline src/cauldrons/CauldronV4.sol]: did you want revert this
```suggestion
        require(!blacklistedCallees[callee], "Cauldron: can't call");
```

File: src/periphery/DegenBoxTokenWrapper.sol [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+/// @notice Wrap token to wrapper and deposit into DegenBox for recipient
+contract DegenBoxTokenWrapper {
+    function wrap(
+        IBentoBoxV1 degenBox,
+        ITokenWrapper wrapper,
+        address recipient,
+        uint256 amount
+    ) external returns (uint256 amountOut, uint256 shareOut) {
+        wrapper.underlying().approve(address(wrapper), amount);
```
Comment (0xm3rlin) [inline src/periphery/DegenBoxTokenWrapper.sol]: couldn't we save this approval by doing it on a constructor level or are you worried that underlying will change?

**Review by 0xm3rlin — APPROVED** [review]
